### PR TITLE
kyria: small changes to allow reducing image size

### PR DESCRIPTION
**Pull request by andresilva**
## Description

This PR makes some small changes to the Kyria keyboard support in order to achieve a smaller image size.
- Move qmk and kyria logo sprites out of `oled_task_kb` so that they can be reused without having to copy the sprite definition.
- Make `oled_task_kb` and `encoder_update_kb` overridable. Users that want to implement their own versions might override these functions to save some space.

## Types of Changes

- [ ] Core
- [ ] Bugfix
- [ ] New feature
- [ ] Enhancement/optimization
- [x] Keyboard (addition or update)
- [ ] Keymap/layout/userspace (addition or update)
- [ ] Documentation

## Checklist

- [x] My code follows the code style of this project: [**C**](https://docs.qmk.fm/#/coding_conventions_c), [**Python**](https://docs.qmk.fm/#/coding_conventions_python)
- [x] I have read the [**PR Checklist** document](https://docs.qmk.fm/#/pr_checklist) and have made the appropriate changes.
- [ ] My change requires a change to the documentation.
- [ ] I have updated the documentation accordingly.
- [x] I have read the [**CONTRIBUTING** document](https://docs.qmk.fm/#/contributing).
- [ ] I have added tests to cover my changes.
- [x] I have tested the changes and verified that they work and don't break anything (as well as I can manage).

### CODE REVIEW

File: keyboards/splitkb/kyria/rev1/rev1.h [hunk 1]
```diff
@@ -20,6 +20,28 @@
 #    include "proton_c.h"
 #endif
 
+#ifdef OLED_ENABLE
+// clang-format off
+static const char PROGMEM qmk_logo[] = {
```
Comment (zvecr) [inline keyboards/splitkb/kyria/rev1/rev1.h:25]: Shouldnt be adding constants to header files, this should be reverted.

File: keyboards/splitkb/kyria/rev1/rev1.c [hunk 2]
```diff
@@ -110,26 +103,14 @@ bool oled_task_kb(void) {
...
-        // clang-format on
         oled_write_raw_P(kyria_logo, sizeof(kyria_logo));
     }
     return false;
 }
 #endif
 
 #ifdef ENCODER_ENABLE
-bool encoder_update_kb(uint8_t index, bool clockwise) {
+__attribute__ ((weak)) bool encoder_update_kb(uint8_t index, bool clockwise) {
```
Comment (zvecr) [inline keyboards/splitkb/kyria/rev1/rev1.c:113]:
```suggestion
bool encoder_update_kb(uint8_t index, bool clockwise) {
```

File: keyboards/splitkb/kyria/rev1/rev1.c [hunk 3]
```diff
@@ -60,19 +60,12 @@ oled_rotation_t oled_init_kb(oled_rotation_t rotation) {
     return OLED_ROTATION_180;
 }
 
-bool oled_task_kb(void) {
+__attribute__ ((weak)) bool oled_task_kb(void) {
```
Comment (zvecr) [inline keyboards/splitkb/kyria/rev1/rev1.c:63]:
```suggestion
bool oled_task_kb(void) {
```

**Review by zvecr — CHANGES_REQUESTED** [review]
> Make oled_task_kb and encoder_update_kb overridable. Users that want to implement their own versions might override these functions to save some space.

Then they should be implementing `_user` functions, and should never try to override a `_kb` function within a keymap.

**Comment by andresilva** [comment]
> > Make oled_task_kb and encoder_update_kb overridable. Users that want to implement their own versions might override these functions to save some space.
> 
> Then they should be implementing `_user` functions, and should never try to override a `_kb` function within a keymap.
 
I understand that users are supposed to use the `_user` functions, and that's what I've used, but I had to override the `_kb` variants to just call the user functions since the defaults include a bunch of unused code (just have a look at the code of both functions I made overridable). This bloats the image size unnecessarily (I only have 28K to play).

File: keyboards/splitkb/kyria/rev1/rev1.h (hunk 1 above)
Comment (andresilva) [reply keyboards/splitkb/kyria/rev1/rev1.h]: Fixed, made it extern and non-static.

**Comment by andresilva** [comment]
> > > Make oled_task_kb and encoder_update_kb overridable. Users that want to implement their own versions might override these functions to save some space.
> > 
> > 
> > Then they should be implementing `_user` functions, and should never try to override a `_kb` function within a keymap.
> 
> I understand that users are supposed to use the `_user` functions, and that's what I've used, but I had to override the `_kb` variants to just call the user functions since the defaults include a bunch of unused code (just have a look at the code of both functions I made overridable). This bloats the image size unnecessarily (I only have 28K to play).

An alternative might be to introduce config variables for this keyboard like `DEFAULT_VISUALIZER` and `DEFAULT_ENCODER` which conditionally compile the default implementations for these components (that I'm trying to override).

File: keyboards/splitkb/kyria/rev1/config.h [hunk 4]
```diff
@@ -85,6 +85,11 @@ along with this program.  If not, see <http://www.gnu.org/licenses/>.
 #ifdef OLED_ENABLE
 #    define OLED_DISPLAY_128X64
 #    define SPLIT_OLED_ENABLE
+#    define DEFAULT_OLED_VISUALIZER
+#endif
+
+#ifdef ENCODER_ENABLE
+#    define DEFAULT_ENCODER_ACTIONS
```
Comment (zvecr) [inline keyboards/splitkb/kyria/rev1/config.h:92]:
```suggestion
```

You should just need to return the correct value from `oled_task_user`, <https://docs.qmk.fm/#/ChangeLog/20211127?id=oled-task-refactor>.

Therefore all `DEFAULT_*` should be reverted.

**Review by zvecr — CHANGES_REQUESTED** [review]

File: keyboards/splitkb/kyria/rev1/config.h (hunk 4 above)
Comment (andresilva) [reply keyboards/splitkb/kyria/rev1/config.h:92]: I feel like my point is not getting across. This is about reducing the amount of code that ends up in the image. If I return false from the user function the code from the kb function will still end up in the image uselessly.

Comment (zvecr) [reply keyboards/splitkb/kyria/rev1/config.h:92]: I understand your end goal of reducing the final firmware size. The issue is with the implementation.

My point is based on previous discussions, this sort of "bespoke" keyboard behaviour isnt one we would want in the repo, nor a paradigm to propagate and later have to unpick.

Comment (andresilva) [reply keyboards/splitkb/kyria/rev1/config.h:92]: That wasn't the case in my test (and I have LTO enabled). I understand that the changes in this PR are non-standard and controversial so I'm OK with just closing this PR and maintaining these changes in my own fork.

**Pull request closed without merging**